// ngx-slimscroll/dist/esm2015/app/ngx-slimscroll/classes/slimscroll-event.class.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function ISlimScrollEvent() { }
function ISlimScrollEvent_tsickle_Closure_declarations() {
    /** @type {?} */
    ISlimScrollEvent.prototype.type;
    /** @type {?|undefined} */
    ISlimScrollEvent.prototype.y;
    /** @type {?|undefined} */
    ISlimScrollEvent.prototype.percent;
    /** @type {?|undefined} */
    ISlimScrollEvent.prototype.duration;
    /** @type {?|undefined} */
    ISlimScrollEvent.prototype.easing;
}
export class SlimScrollEvent {
    /**
     * @param {?=} obj
     */
    constructor(obj) {
        this.type = obj.type;
        this.y = obj && obj.y ? obj.y : 0;
        this.percent = obj && obj.percent ? obj.percent : 0;
        this.duration = obj && obj.duration ? obj.duration : 0;
        this.easing = obj && obj.easing ? obj.easing : 'linear';
    }
}
function SlimScrollEvent_tsickle_Closure_declarations() {
    /** @type {?} */
    SlimScrollEvent.prototype.type;
    /** @type {?} */
    SlimScrollEvent.prototype.y;
    /** @type {?} */
    SlimScrollEvent.prototype.percent;
    /** @type {?} */
    SlimScrollEvent.prototype.duration;
    /** @type {?} */
    SlimScrollEvent.prototype.easing;
}
