// @angular/compiler-cli/src/ngtsc/compiler_host.js
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define("@angular/compiler-cli/src/ngtsc/compiler_host", ["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    /**
     * Implementation of `CompilerHost` which delegates to a native TypeScript host in most cases.
     */
    var NgtscCompilerHost = /** @class */ (function () {
        function NgtscCompilerHost(delegate) {
            this.delegate = delegate;
            if (delegate.resolveTypeReferenceDirectives) {
                this.resolveTypeReferenceDirectives = function (names, containingFile) {
                    return delegate.resolveTypeReferenceDirectives(names, containingFile);
                };
            }
        }
        NgtscCompilerHost.prototype.getSourceFile = function (fileName, languageVersion, onError, shouldCreateNewSourceFile) {
            return this.delegate.getSourceFile(fileName, languageVersion, onError, shouldCreateNewSourceFile);
        };
        NgtscCompilerHost.prototype.getDefaultLibFileName = function (options) {
            return this.delegate.getDefaultLibFileName(options);
        };
        NgtscCompilerHost.prototype.writeFile = function (fileName, data, writeByteOrderMark, onError, sourceFiles) {
            return this.delegate.writeFile(fileName, data, writeByteOrderMark, onError, sourceFiles);
        };
        NgtscCompilerHost.prototype.getCurrentDirectory = function () { return this.delegate.getCurrentDirectory(); };
        NgtscCompilerHost.prototype.getDirectories = function (path) { return this.delegate.getDirectories(path); };
        NgtscCompilerHost.prototype.getCanonicalFileName = function (fileName) {
            return this.delegate.getCanonicalFileName(fileName);
        };
        NgtscCompilerHost.prototype.useCaseSensitiveFileNames = function () { return this.delegate.useCaseSensitiveFileNames(); };
        NgtscCompilerHost.prototype.getNewLine = function () { return this.delegate.getNewLine(); };
        NgtscCompilerHost.prototype.fileExists = function (fileName) { return this.delegate.fileExists(fileName); };
        NgtscCompilerHost.prototype.readFile = function (fileName) { return this.delegate.readFile(fileName); };
        NgtscCompilerHost.prototype.loadResource = function (file) { throw new Error('Method not implemented.'); };
        NgtscCompilerHost.prototype.preloadResource = function (file) { throw new Error('Method not implemented.'); };
        return NgtscCompilerHost;
    }());
    exports.NgtscCompilerHost = NgtscCompilerHost;
});
