// @rxweb/reactive-form-validators/esm5/util/credit-card-regex.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var CreditCardRegex = /** @class */ (function () {
    function CreditCardRegex() {
        this.Visa = new RegExp('^(?:4[0-9]{12})(?:[0-9]{3})?$');
        this.AmericanExpress = new RegExp('^(?:3[47][0-9]{13})$');
        this.Maestro = new RegExp('^(?:(?:5[0678]\\d\\d|6304|6390|67\\d\\d)\\d{8,15})$');
        this.JCB = new RegExp('^(?:(?:2131|1800|35\\d{3})\\d{11})$');
        this.Discover = new RegExp('^(?:6(?:011|5[0-9]{2})(?:[0-9]{12}))$');
        this.DinersClub = new RegExp('^(?:3(?:0[0-5]|[68][0-9])[0-9]{11})$');
        this.MasterCard = new RegExp('^(?:5[1-5][0-9]{2}|222[1-9]|22[3-9][0-9]|2[3-6][0-9]{2}|27[01][0-9]|2720)[0-9]{12}$');
    }
    return CreditCardRegex;
}());
export { CreditCardRegex };
function CreditCardRegex_tsickle_Closure_declarations() {
    /** @type {?} */
    CreditCardRegex.prototype.Visa;
    /** @type {?} */
    CreditCardRegex.prototype.AmericanExpress;
    /** @type {?} */
    CreditCardRegex.prototype.Maestro;
    /** @type {?} */
    CreditCardRegex.prototype.JCB;
    /** @type {?} */
    CreditCardRegex.prototype.Discover;
    /** @type {?} */
    CreditCardRegex.prototype.DinersClub;
    /** @type {?} */
    CreditCardRegex.prototype.MasterCard;
}
