// ngx-highlightjs/esm2015/lib/highlight.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function HighlightOptions() { }
function HighlightOptions_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    HighlightOptions.prototype.theme;
    /** @type {?|undefined} */
    HighlightOptions.prototype.path;
    /** @type {?|undefined} */
    HighlightOptions.prototype.auto;
    /** @type {?|undefined} */
    HighlightOptions.prototype.config;
}
/**
 * @record
 */
export function HighlightConfig() { }
function HighlightConfig_tsickle_Closure_declarations() {
    /**
     * tabReplace: a string used to replace TAB characters in indentation.
     * @type {?|undefined}
     */
    HighlightConfig.prototype.tabReplace;
    /**
     * useBR: a flag to generate <br> tags instead of new-line characters in the output, useful when code is marked up using a non-<pre> container.
     * @type {?|undefined}
     */
    HighlightConfig.prototype.useBR;
    /**
     * classPrefix: a string prefix added before class names in the generated markup, used for backwards compatibility with stylesheets.
     * @type {?|undefined}
     */
    HighlightConfig.prototype.classPrefix;
    /**
     * languages: an array of language names and aliases restricting auto detection to only these languages.
     * @type {?|undefined}
     */
    HighlightConfig.prototype.languages;
}
/**
 * @record
 */
export function HighlightResult() { }
function HighlightResult_tsickle_Closure_declarations() {
    /** @type {?|undefined} */
    HighlightResult.prototype.language;
    /** @type {?|undefined} */
    HighlightResult.prototype.r;
    /** @type {?|undefined} */
    HighlightResult.prototype.second_best;
    /** @type {?|undefined} */
    HighlightResult.prototype.top;
    /** @type {?|undefined} */
    HighlightResult.prototype.value;
}
