// ngx-slimscroll/dist/esm5/app/ngx-slimscroll/classes/slimscroll-state.class.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function ISlimScrollState() { }
function ISlimScrollState_tsickle_Closure_declarations() {
    /** @type {?} */
    ISlimScrollState.prototype.scrollPosition;
    /** @type {?} */
    ISlimScrollState.prototype.isScrollAtStart;
    /** @type {?} */
    ISlimScrollState.prototype.isScrollAtEnd;
}
var SlimScrollState = /** @class */ (function () {
    function SlimScrollState(obj) {
        this.scrollPosition = obj && obj.scrollPosition ? obj.scrollPosition : 0;
        this.isScrollAtStart = obj && typeof obj.isScrollAtStart !== 'undefined' ? obj.isScrollAtStart : true;
        this.isScrollAtEnd = obj && typeof obj.isScrollAtEnd !== 'undefined' ? obj.isScrollAtEnd : false;
    }
    return SlimScrollState;
}());
export { SlimScrollState };
function SlimScrollState_tsickle_Closure_declarations() {
    /** @type {?} */
    SlimScrollState.prototype.scrollPosition;
    /** @type {?} */
    SlimScrollState.prototype.isScrollAtStart;
    /** @type {?} */
    SlimScrollState.prototype.isScrollAtEnd;
}
