// @rxweb/reactive-form-validators/esm5/core/validator.interface.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function DecoratorConfiguration() { }
function DecoratorConfiguration_tsickle_Closure_declarations() {
    /** @type {?} */
    DecoratorConfiguration.prototype.annotationType;
    /** @type {?} */
    DecoratorConfiguration.prototype.propertyName;
    /** @type {?} */
    DecoratorConfiguration.prototype.propertyIndex;
    /** @type {?|undefined} */
    DecoratorConfiguration.prototype.target;
    /** @type {?|undefined} */
    DecoratorConfiguration.prototype.propertyConstructor;
    /** @type {?|undefined} */
    DecoratorConfiguration.prototype.config;
}
/**
 * @record
 */
export function InstanceContainer() { }
function InstanceContainer_tsickle_Closure_declarations() {
    /** @type {?} */
    InstanceContainer.prototype.instance;
    /** @type {?} */
    InstanceContainer.prototype.propertyAnnotations;
    /** @type {?|undefined} */
    InstanceContainer.prototype.properties;
    /** @type {?|undefined} */
    InstanceContainer.prototype.conditionalValidationProps;
    /** @type {?|undefined} */
    InstanceContainer.prototype.conditionalObjectProps;
}
/**
 * @record
 */
export function PropertyInfo() { }
function PropertyInfo_tsickle_Closure_declarations() {
    /** @type {?} */
    PropertyInfo.prototype.name;
    /** @type {?} */
    PropertyInfo.prototype.propertyType;
    /** @type {?|undefined} */
    PropertyInfo.prototype.entity;
}
